build_resolvers/lib/src/build_asset_uri_resolver.dart

```diff
@@ -27,54 +39,47 @@ class BuildAssetUriResolver extends UriResolver {
   /// later). If this happens we don't want to hide the asset from the analyzer.
   final seenAssets = Set<AssetId>();
 
-  Future<void> performResolve(
-      BuildStep buildStep, List<AssetId> entryPoints, AnalysisDriver driver) {
-    // Basic approach is to start at the first file, update it's contents
-    // and see if it changed, then walk all files accessed by it.
-    var visited = Set<AssetId>();
-    var visiting = _FutureGroup();
-
-    void processAsset(AssetId assetId) {
-      visited.add(assetId);
-
-      visiting.add(buildStep.readAsString(assetId).then((contents) {
-        if (_cachedAssetContents[assetId] != contents) {
-          if (_cachedAssetContents.containsKey(assetId)) {
-            var path = assetPath(assetId);
-            resourceProvider.updateFile(path, contents);
-            driver.changeFile(path);
-          } else {
-            resourceProvider.newFile(assetPath(assetId), contents);
-          }
-          _cachedAssetContents[assetId] = contents;
-          var unit = parseDirectives(contents, suppressErrors: true);
-          var dependencies = unit.directives
-              .whereType<UriBasedDirective>()
-              .where((directive) {
-                var uri = Uri.parse(directive.uri.stringValue);
-                return !_ignoredSchemes.any(uri.isScheme);
-              })
-              .map((d) => AssetId.resolve(d.uri.stringValue, from: assetId))
-              .where((id) => id != null)
-              .toSet();
-          _cachedAssetDependencies[assetId] = dependencies;
+  Future<void> performResolve(BuildStep buildStep, List<AssetId> entryPoints,
+      AnalysisDriver driver) async {
+    final states =
+        await crawlAsync<AssetId, _AssetState>(entryPoints, (id) async {
+      try {
+        return _AssetState(id, await buildStep.readAsString(id));
+      } catch (e) {
+        if (seenAssets.contains(id)) {
+          // ignore from this graph, some later build step may still be using it
+          // so it shouldn't be removed from [resourceProvider], but we also
+          // don't care about it's transitive imports.
+          return null;
         }
-        _cachedAssetDependencies[assetId]
-            .where((id) => !visited.contains(id))
-            .forEach(processAsset);
-      }, onError: (e) {
-        if (seenAssets.contains(assetId)) return;
-        _cachedAssetDependencies.remove(assetId);
-        _cachedAssetContents.remove(assetId);
-        final path = assetPath(assetId);
+        _cachedAssetDependencies.remove(id);
+        _cachedAssetContents.remove(id);
+        return _AssetState.removed(id);
+      }
+    }, (id, state) {
+      if (state.isRemoved) return const [];
+      if (_cachedAssetContents[id] == state.content) {
+        return _cachedAssetDependencies[id];
+      }
+      _cachedAssetContents[id] = state.content;
+      return _cachedAssetDependencies[id] = _parseDirectives(state);
+    }).toList();
```

Comment:
Can we leave this as a stream and use await for, or does that bring back the bug? I am mostly concerned about the potential memory consumption of this list (it could get large and we could have a lot of them).

Reply:
That seems to restore the problem.

Reply:
I noticed an optimization to bring us closer to previous behavior - I was doing updates here even for assets that hadn't changed. If I add another `isChanged` boolean I can filter the list down to just the changes which should help keep it smaller.

Reply:
for reasons I don't understand it turns out I _must_ use `await for` and collect up the changed paths to update the `driver` afterwards. I have no idea why updating the `resourceProvider` must happen during the async crawling...

I also don't know why the problem doesn't surface in this repo but does show up when I try to build a larger project...

Reply:
:/

Reply:
Hmm, we've got a much worse bug here than I thought.

Somehow all the way back in https://github.com/dart-lang/build/pull/1983 I was never writing to `seenAssets`, so all of our out-of-order build step handling was completely broken. I don't know how this hasn't been manifesting in even worse ways...

Reply:
OK, there are some really subtle async things happening. I think what I'm seeing is that we can't have any async steps between updating `_cachedAsset*` and the `resourceProvider`.